unified-runtime/source/adapters/opencl/usm.hpp

```diff
@@ -17,44 +17,59 @@
 //
 // Example usage:
 //
-// auto Info = new AllocDeleterCallbackInfo(USMFreeFuncPtr, Context,
+// auto Info = new AllocDeleterCallbackInfoIntel(USMFreeFuncPtr, Context,
 // Allocation); clSetEventCallback(USMOpEvent, CL_COMPLETE,
 // AllocDeleterCallback, Info);
-struct AllocDeleterCallbackInfo {
-  AllocDeleterCallbackInfo(clMemBlockingFreeINTEL_fn USMFree,
-                           cl_context CLContext, void *Allocation)
-      : USMFree(USMFree), CLContext(CLContext), Allocation(Allocation) {
+
+struct AllocDeleterCallbackInfoBase {
+  AllocDeleterCallbackInfoBase(cl_context Context, uint8_t *Allocation)
+      : CLContext(Context), Allocation(Allocation) {
     clRetainContext(CLContext);
   }
-  ~AllocDeleterCallbackInfo() {
-    USMFree(CLContext, Allocation);
-    clReleaseContext(CLContext);
-  }
-  AllocDeleterCallbackInfo(const AllocDeleterCallbackInfo &) = delete;
-  AllocDeleterCallbackInfo &
-  operator=(const AllocDeleterCallbackInfo &) = delete;
 
-  clMemBlockingFreeINTEL_fn USMFree;
+  virtual ~AllocDeleterCallbackInfoBase() = 0;
+
+  AllocDeleterCallbackInfoBase(const AllocDeleterCallbackInfoBase &) = delete;
+  AllocDeleterCallbackInfoBase &
+  operator=(const AllocDeleterCallbackInfoBase &) = delete;
+
+protected:
   cl_context CLContext;
-  void *Allocation;
+  uint8_t *Allocation;
+};
+
+inline AllocDeleterCallbackInfoBase::~AllocDeleterCallbackInfoBase() {
+  clReleaseContext(CLContext);
+}
+
+struct AllocDeleterCallbackInfo : AllocDeleterCallbackInfoBase {
+  AllocDeleterCallbackInfo(cl_context CLContext, uint8_t *Allocation)
+      : AllocDeleterCallbackInfoBase(CLContext, Allocation) {}
+
+  ~AllocDeleterCallbackInfo() override { delete[] Allocation; }
 };
 
-struct AllocDeleterCallbackInfoWithQueue : AllocDeleterCallbackInfo {
+struct AllocDeleterCallbackInfoIntel : AllocDeleterCallbackInfoBase {
```

Comment:
I think something like AllocDeleterCallbackInfoUSM is better, these might not be intel extension functions forever